git: Add ability to pass `--signoff`

ericcornelissen opened the pull request:
This adds an option for `--signoff` to the git panel and commit modal. It allows users to enable the [`--signoff` flag](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt-code--signoffcode) when committing through Zed. The option is added to the context menu of the commit button (following the style of the "Editor Controls").

To support this, the commit+amend experience was revamped (following the ideas of [this comment](https://github.com/zed-industries/zed/pull/29874#issuecomment-2950848000)). Amending is now also a toggle in the commit button's dropdown menu. I've kept some of the original experience such as the changed button text and ability to cancel outside the context menu.

The tooltip of the commit buttons now also includes the flags that will be used based on the amending and signoff status (which I couldn't capture in screenshots unfortunately). So, by default the tooltip will say `git commit` and if you toggle, e.g., amending on it will say `git commit --amend`.

| What | Panel | Modal |
| --- | --- | --- |
| Not amending, dropdown | ![git modal preview, not amending, dropdown](https://github.com/user-attachments/assets/82c2b338-b3b5-418c-97bf-98c33202d7dd) | ![commit modal preview, not amending, dropdown](https://github.com/user-attachments/assets/f7a6f2fb-902d-447d-a473-2efb4ba0f444) |
| Amending, dropdown | ![git modal preview, amending, dropdown](https://github.com/user-attachments/assets/9e755975-4a27-43f0-aa62-be002ecd3a92) | ![commit modal preview, amending, dropdown](https://github.com/user-attachments/assets/cad03817-14e1-46f6-ba39-8ccc7dd12161) |
| Amending | ![git modal preview, amending](https://github.com/user-attachments/assets/e1ec4eba-174e-4e5f-9659-5867d6b0fdc2) | - |

The initial implementation was based on the changeset of https://github.com/zed-industries/zed/pull/28187.

Closes https://github.com/zed-industries/zed/discussions/26114

Release Notes:

- Added git `--signoff` support.
- Update the git `--amend` experience.
- Improved git panel to persist width as well as amend and signoff on a per-workspace basis.

ericcornelissen commented:
@cla-bot check

mikayla-maki commented:
Hmmm, I don't think I like adding a niche git feature to the always-visible UI.... what do you think about putting it in the commit dropdown, kind of like we do with the visual editor settings? 

<img width="254" alt="Screenshot 2025-05-06 at 6 35 11 PM" src="https://github.com/user-attachments/assets/1f345966-7079-49f0-b956-d7b67e92802d" />

This might be nice for automatically generating a `--no-signoff` as well, if you're doing an amend and deselect the option.

4ad commented:
This is not a niche git feature, this is an essential git feature for working on certain projects. Making it inconvenient to set defeats the point of this feature, and in turn makes Zed inconvenient. Additionally if it's not visible, then it's very easy to forget to set it (and it has to be set each explicitly time, for each commit, set and forget is not the right behavior, because it require an explicit action each time you commit).

FWIW, JetBrains has this checkbox exactly in the same place in their git dialog.

ericcornelissen commented:
> This is not a niche git feature, this is an essential git feature for working on certain projects. Making it inconvenient to set defeats the point of this feature, and in turn makes Zed inconvenient. Additionally if it's not visible, then it's very easy to forget to set it (and it has to be set each explicitly time, for each commit, set and forget is not the right behavior, because it require an explicit action each time you commit).

I think that's fair, my interpretation of @mikayla-maki's comment to make it a toggle in the menu so that you only have to set it once. I haven't seen/used per-project configurations in Zed (only recently started using it) but I agree it would be ideal if the toggle was enabled on a per-project basis.

> FWIW, JetBrains has this checkbox exactly in the same place in their git dialog.

FWIW VSCode has it as an editor-wide setting or per-project setting if you set it in `.vscode/settings.json`.

mikayla-maki commented:
@ericcornelissen The exact semantics of whether you have to select it each time, or can have it pre-set by settings are up to y'all to decide. I was thinking of something that works more like an additional option next to amend in the purple area below.  That said, I still think the presence of this checkbox clutters the UI somewhat, especially as we already have several other similar buttons in that location, note red box with the the "generate commit message" magic button and the "add automatic co-authors" button in that space. If we added the checkbox above, we'd have even more options, each of which work totally differently.

<img width="623" alt="Screenshot 2025-05-08 at 5 50 36 PM" src="https://github.com/user-attachments/assets/9014b6db-249d-4402-ae1c-05774a7823bb" />

However, I see where @4ad is coming from, particularly the part about how JetBrains already includes this checkbox by default. What do y'all think about adding a setting, e.g. `"git": { "enable-sign-offs": true/false }`, that could be set in the per-project settings, that controls whether this checkbox shows up in that area?

ericcornelissen commented:
@mikayla-maki would it make sense to do both? That is, expand the commit dropdown menu with a "Signoff" option as well as add an option to add the button next to the branch/generate message/co-authors buttons? I personally think it does as the menu makes the option always available, and it would also be more scalable to a larger number of git commit options. The buttons could then be enabled by users that want to use a specific option more often. What do you think? (Else I'd personally be in favor of the dropdown.)

For the context menu, I still feel that a permanent (per-project) toggle makes the most sense - otherwise it would be easy to miss. Having started to implement this, I quickly realized it would need quite a significant overhaul of the way the amend option is currently implemented. This is because otherwise the other commit options are not available when you're amending. I personally feel it would suffice to change the existing "Amend" option to a toggle as well (keeping the _"This will update your most recent commit."_ message). Would that work?

For the checkbox in the button area, should this be done using an icon instead of a checkbox+label? If so, I imagine that's subject to the [requirements for file icons](https://github.com/zed-industries/zed/blob/b34f19a46f3be515d324d5ec21c0f0e19e941794/CONTRIBUTING.md#file-icons)? Since I'm not much of a designer myself, I'd need some help with that. Also, I'm not sure how to visualize signoff, an icon of a signature is more likely to come off as enabling [commit signatures (i.e. `-S`)](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt-code-Sltkey-idgtcode).

cole-miller commented:
Thanks for working on this @ericcornelissen and sorry for the long delay here.

I think we should add a toggleable "Sign off" entry to the context menu on the commit button, and a project setting that controls whether this is checked by default or not. (The "Sign off" entry's state should probably reset to this default whenever you commit.) It seems to me that typically within a given repo you'll either want this always enabled or always disabled, so if we have that setting I don't think we need to add a button to make it possible to toggle without opening the context menu.

I agree with your proposal to rework `Amend` to be toggleable as well!

ericcornelissen commented:
@cole-miller thanks for the response and apologies for the delay in my end this time around. I'll look into updating this in the coming ~week.

ericcornelissen committed:
Add `--signoff` option

Add a toggleable entry to the commit button dropdown in the to the git
panel and commit modal which can be used to enable signing off on git
commits. This allows users to enable the `--signoff` flag when
committing through Zed.

To support this, the `--amend` option experience has been reworked. It
now also works as a toggleable entry in the dropdown, no longer changing
the commit button (apart from its text). While it is possible to disable
the amend from the dropdown, the ability to cancel without the dropdown
has been maintained.

ericcornelissen commented:
@cole-miller I pushed an update - rebased from 46834d31f139b7d9d1fdf1a1e072437f7f128e29 - that implements most of the functionality (the Pull Request description is updated accordingly) except for looking up the default value from the config (though I did manage to get its state to persist after a restart at least, I hope to be able to look at that by myself).

I'll try to book a time to pair for next week to go through the changes and fix some stuff up as I'm sure my implementation has some rough edges. If nothing else there's some unchanged details I'm unsure of..

ericcornelissen committed:
Serialize git panel per project

Based on f98c6fb2cf86e16de6860061c2ae33b6c6d5b5ab, this changes the
serialization of the git panel from global to per-workspace. This
follows the addition of the signoff option to the git panel, meaning
this option is persistes per-workspace. We take advantage of this change
to also persist the amend pending state for workspaces. Moreover, this
aligns the behavior of the git panel (specifically its width) with that
of the other panels.

cole-miller reviewed (approved):
Thanks @ericcornelissen, great work on this!

SomeoneToIgnore commented:
Heads-up, this broke git panel serialization (https://github.com/zed-industries/zed/issues/34618 and https://github.com/zed-industries/zed/issues/34611) and that particular code is somewhat odd to run without the debounce.

https://github.com/zed-industries/zed/pull/34622 follow-up fixes this.
